service/qos/service_level_controller.cc

```diff
@@ -231,15 +231,30 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
 }
 
 future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {
-    _service_levels_db.emplace(name, sl_data);
-    return make_ready_future();
+    return seastar::async( [this, name, sl_data] {
+        _subscribers.for_each([name, sl_data] (qos_configuration_change_subscriber* subscriber) {
+            subscriber->on_before_service_level_add(name, sl_data.slo).get();
+        });
+    }).then([this, name, sl_data] () {
+        _service_levels_db.emplace(name, sl_data);
+        return make_ready_future();
+    });
+
 }
 
 future<> service_level_controller::notify_service_level_updated(sstring name, service_level_options slo) {
     auto sl_it = _service_levels_db.find(name);
     future<> f = make_ready_future();
     if (sl_it != _service_levels_db.end()) {
-        sl_it->second.slo = slo;
+        service_level_options slo_before = sl_it->second.slo;
+        return seastar::async( [this, name, slo_before, slo] {
+            _subscribers.for_each([name, slo_before, slo] (qos_configuration_change_subscriber* subscriber) {
+                subscriber->on_before_service_level_change(name, slo_before, slo).get();
+            });
+        }).then([this, sl_it, slo] () {
```

Comment:
ditto - let's move contents of this continuation into the thread, since one was already created